### PR TITLE
Improved feedback

### PR DESCRIPTION
**Pull request by hughdickinson**
Staging branch URL: https://pr-5475.pfe-preview.zooniverse.org

Enhances experimental realtime feedback. New functionality:

* Identical feedback messages corresponding to multiple feedback marks can be pluralized using the special `$count}` placeholder.  This is configurable in the project builder. See `app/features/feedback/README.md`.
* For the radial tool, the feedback mark colors are customizable (independently for success and failure cases). This is configurable in the project builder. 
* For the radial tool, the feedback mark opacity is customizable. This is configurable in the project builder. 
* For the radial tool different shapes can be used to mark successes and failures. This is configurable in the project builder. 

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [x] Does talk load correctly?
- [x] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

### CODE REVIEW

**Comment by hughdickinson** [comment]
> Enhances experimental realtime feedback. New functionality:
> 
> > Identical feedback messages corresponding to multiple feedback marks can be pluralized using the special $count} placeholder. This is configurable in the project builder. See app/features/feedback/README.md.
> 
> I'm not sure if it is the right approach to add another configurable message using subject metadata. This is adding more complexity to the configuration that some project builders have already struggled with. In the rewrite, we add a reduced count of the number of times a success or fail happens next to its corresponding message. See: [zooniverse/front-end-monorepo#892](https://github.com/zooniverse/front-end-monorepo/pull/892)
> 
> I think it is a good idea to have a reduced count, but I think we should consistent with what we're doing in the rewrite. Do you have a reasoning as to why this might be a better approach? Does this approach outweigh the difficulty project owners may have in configuration?

My motivation for implementing it this way (with a plural message and a singular message) was to allow for cases when there can be one or more successes or failures. In that case, providing a meaningful message may involve pluralization of nouns as well as just adding a number e.g. 

"You found an object" becomes "You found 10 objects". 

This allows more natural sounding feedback and I imagine it would also be more flexible for different languages although I'm not certain about that.
 
> > For the radial tool, the feedback mark colors are customizable (independently for success and failure cases). This is configurable in the project builder. For the radial tool, the feedback mark opacity is customizable. This is configurable in the project builder.
>
> We generally don't allow style customizations by project owners. Could you explain the rational for allowing this?

There are two reasons.
1. Hard-coded colours may be invisible on certain subjects.
2. The default color scheme in this case was red/green which is not great for anyone with red-green colour blindness.

> > For the radial tool different shapes can be used to mark successes and failures. This is configurable in the project builder.
> 
> Again this is adding configuration complexity that may be difficult for project builders to understand and configure. This is a good idea, however, I think we can just have a non-configurable general solution for this.

At the moment this is the least configurable of all the changes. It's just a simple on/off toggle. I can imagine situations for which users may not want to show squares for failures on the radial feedback tool, but for my purposes I'd be happy to hard code it to be on.

> Is there a project we can manually test these enhancements with?

I had expected the branch to be automatically staged when I submitted the PR, but it seems not to have been. Will this happen? I can manually stage it if not. Once it is staged, this project/workflow demonstrates the functionality

`/hughdickinson/galaxy-zoo-clump-scout/classify?workflow=11223`

File: app/features/feedback/README.md [hunk 1]
```diff
@@ -14,15 +15,15 @@
 ### Workflow
 
 1. In the project builder, navigate to your workflow, and find the feedback section in the bottom right.
-1. Start creating a new rule by clicking the button.
-1. Define a unique id for the rule, check the boxes for whether you want feedback to be shown on success, failure or both. Where enabled, a default message must be defined.
-1. Select a strategy for your rule. See the Strategies section below for more information.
-1. Define any additional options required by the chosen strategy, and save the new rule.
+2. Start creating a new rule by clicking the button.
```
Comment (eatyourgreens) [inline app/features/feedback/README.md]: These numbers are calculated by the Markdown parser. The original code was the markdown for a numbered list.

File: app/features/feedback/lab/components/feedback-section.jsx [hunk 2]
```diff
@@ -12,61 +17,147 @@ counterpart.registerTranslations('en', {
...
+
+        <CheckboxInput title={fieldText('pluralFailureMessagesEnabled.title')} help={fieldText('pluralFailureMessagesEnabled.help')} name="pluralFailureMessagesEnabled" checked={this.state.form.pluralFailureMessagesEnabled} onChange={this.handleInputChange.bind(this)}/>
+
+        <CheckboxInput title={fieldText('colorizeUniqueMessagesEnabled.title')} help={fieldText('colorizeUniqueMessagesEnabled.help')} name="colorizeUniqueMessagesEnabled" checked={this.state.form.colorizeUniqueMessagesEnabled} onChange={this.handleInputChange.bind(this)}/> {this.state.form.colorizeUniqueMessagesEnabled && <SelectInput title={fieldText('allowedSuccessFeedbackMarkerColors.title')} help={fieldText('allowedSuccessFeedbackMarkerColors.help')} options={colorOptions} name="allowedSuccessFeedbackMarkerColors" multi={true} value={this.state.allowedSuccessFeedbackMarkerColors} placeholder={fieldText('allowedSuccessFeedbackMarkerColors.placeholder')} onChange={this.handleAllowedSuccessFeedbackMarkerColors.bind(this)}/>}
```
Comment (eatyourgreens) [inline app/features/feedback/lab/components/feedback-section.jsx]: Long lines like this should be flagged by eslint. Can you lint your changes, please?

File: app/features/feedback/lab/containers/feedback-section-container.jsx [hunk 3]
```diff
@@ -42,8 +44,22 @@ class FeedbackSectionContainer extends Component {
         rules: []
       }
     });
-
+    console.log(editedTask.rules);
```
Comment (eatyourgreens) [inline app/features/feedback/lab/containers/feedback-section-container.jsx]: Stray console log.

File: app/features/feedback/lab/containers/feedback-section-container.jsx [hunk 4]
```diff
@@ -42,8 +44,22 @@ class FeedbackSectionContainer extends Component {
...
     editedTask.feedback.rules = _.unionBy([rule], editedTask.feedback.rules, 'id');
+    console.log(editedTask.feedback.rules);
```
Comment (eatyourgreens) [inline app/features/feedback/lab/containers/feedback-section-container.jsx]: Stray console log.

File: app/features/feedback/shared/helpers/get-global-feedback-options-from-rules.jsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  }).value();
+  const colorizeUniqueMessagesEnabled = _.chain(rules).some(rule => {
+    return rule.colorizeUniqueMessagesEnabled;
+  }).value();
+  const successFailureShapesEnabled = _.chain(rules).some(rule => {
+    return rule.successFailureShapesEnabled;
+  }).value();
+  const allowedSuccessFeedbackMarkerColors = _.chain(rules).some().value() ? rules[0].allowedSuccessFeedbackMarkerColors : getPfeMarkerColors();
+  const allowedFailureFeedbackMarkerColors = _.chain(rules).some().value() ? rules[0].allowedFailureFeedbackMarkerColors : getPfeMarkerColors();
+  return {pluralSuccessMessagesEnabled: pluralSuccessMessagesEnabled, pluralFailureMessagesEnabled: pluralFailureMessagesEnabled, colorizeUniqueMessagesEnabled: colorizeUniqueMessagesEnabled, successFailureShapesEnabled: successFailureShapesEnabled, allowedSuccessFeedbackMarkerColors: allowedSuccessFeedbackMarkerColors, allowedFailureFeedbackMarkerColors: allowedFailureFeedbackMarkerColors};
```
Comment (eatyourgreens) [inline app/features/feedback/shared/helpers/get-global-feedback-options-from-rules.jsx]: This needs linting too. Redundant property declarations in objects should be caught by the linter.

File: package-lock.json [hunk 6]
```diff
@@ -9,16 +9,16 @@
...
-        "commander": "^2.8.1",
-        "convert-source-map": "^1.1.0",
-        "fs-readdir-recursive": "^1.1.0",
-        "glob": "^7.0.0",
-        "lodash": "^4.17.10",
-        "mkdirp": "^0.5.1",
-        "output-file-sync": "^2.0.0",
-        "slash": "^2.0.0",
-        "source-map": "^0.5.0"
+        "chokidar": "2.0.4",
```
Comment (eatyourgreens) [inline package-lock.json]: This PR doesn't install or update any packages, so it shouldn't be changing the package lock file.

**Comment by eatyourgreens** [comment]
As a volunteer on the test workflow, how do I know which types of clump I've missed? I'm seeing red squares and a message along the lines of `you missed 3 clumps` but I don't know which of the two types of drawing tool this refers to.

**Comment by eatyourgreens** [comment]
For anyone else who'd like to test this, the link is https://pr-5475.pfe-preview.zooniverse.org/projects//hughdickinson/galaxy-zoo-clump-scout/classify?workflow=11223&env=production and you must be in admin mode.

**Comment by eatyourgreens** [comment]
Yellow feedback is impossible to read on a white background.
Screenshot showing a feedback modal with yellow text on a white background.

**Comment by eatyourgreens** [comment]
My initial reaction is that this is an interesting idea but it needs more work before it's ready to be used on a project. Ideally, it should be added to feedback in [the new classifier](https://github.com/zooniverse/front-end-monorepo/tree/master/packages/lib-classifier), since we are moving projects across to using that code.

**Comment by srallen** [comment]
> My motivation for implementing it this way (with a plural message and a singular message) was to allow for cases when there can be one or more successes or failures. In that case, providing a meaningful message may involve pluralization of nouns as well as just adding a number

I think the idea to support interpolation is a good one, but I don't think we should add yet another subject metadata field just for pluralization support. We can support this in the default messaging fields for rules that can be setup. Since the message will be the same, the default message can be used rather than use the subject metadata field messaging. Pretty much all the projects that have used feedback so far have misunderstood the subject metadata field for the message as being required when it's not. I'm afraid this would happen with this new field as well.

> Hard-coded colours may be invisible on certain subjects.
> The default color scheme in this case was red/green which is not great for anyone with red-green colour blindness.

The onus for good accessible design is on us, not the project builders. Allowing the colors to be customizable doesn't actually solve the accessibility problem because we're solely using color to communicate information. What the change actually should be is using color plus something else to indicate success or failure like icons or keys. This was discussed in https://github.com/zooniverse/Panoptes-Front-End/issues/5332

**Comment by eatyourgreens** [comment]
I think I've got an issue (or a comment on an issue?) somewhere where I outline how string interpolation might work for feedback. I can't find it, though. It might have been a suggestion for the new classifier.

**Comment by srallen** [comment]
Yep, we talked about it here: https://github.com/zooniverse/front-end-monorepo/issues/774

**Comment by hughdickinson** [comment]
Thanks all for the feedback. I agree that doing this properly requires a lot of careful work. Ultimately, since the MVP for our project is that multiple identical messages get reduced, I'll close this PR and open a new one that *only* performs the reduction and renders the messages in the same way as the current monorepo example (with the repeat count in parentheses after the message).

**Pull request closed without merging**